Feature/follow conversation tooltip

alpavanoglu opened the pull request:
Refs pbArwn-4oo-p2

To test:
1. Enable local feature flag "Feature Highlight Tooltip".
2. Go to Reader, into any post.
3. A "New" anchor should appear (fade in) if the "Follow Conversation" button is not visible due to content size.
4. Tapping on it should scroll to the "Follow Conversation" button and the tooltip should animate in.
5. If "Follow Conversation" button is immediately visible, then anchor shouldn't appear as the button is already there.
6. The feature shouldn't be visible after the following: 
  - "Got it" button is tapped.
  - The tooltip is displayed 3 times without being interacted.
  - When the post is already subscribed/followed. 
  
The following snippet can be added to `ReaderDetailsViewController:viewDidLoad` to reset the local memory.
  ```Swift
          // FIXME: Remove, temporary code
        FeatureHighlightStore.didDismissTooltip = false
        FeatureHighlightStore.followConversationTooltipCounter = 0
  ```
  
Follow Up Points:
1. Sparkle emoji will be replaced with an asset for accessibility.
3. The tooltip should be dismissed when "Follow Conversation" button is tapped.
4. Add tracking
5. Refactor `FeatureHighlightStore` to make it testable and add unit tests.
6. Remove local feature flag

## Regression Notes
1. Potential unintended areas of impact
The additions aren't intrusive. There shouldn't be any regression.

8. What I did to test those areas of impact (or what existing automated tests I relied on)
`FeatureHighlightStore` will be refactored and unit tested in a follow up PR. The rest is UI code coupled to `ReaderDetailsViewController`. A big refactoring is needed to make things testable on that VC.

9. What automated tests I added (or what prevented me from doing so)
See 8.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

alpavanoglu committed:
Add tooltip for Follow Conversations

Add `TooltipVerticalPosition`

alpavanoglu committed:
Add scroll on anchor interaction behavior

alpavanoglu committed:
Update `TooltipPresenter` to support target point

alpavanoglu committed:
Move tooltip from header to view controller

alpavanoglu committed:
Add `showTooltip` to `likesUpdated`

alpavanoglu committed:
Update anchor visibility

alpavanoglu committed:
Finalize scroll, display and hiding behavior

alpavanoglu committed:
Add strings and fade-in/out animation for anchor

alpavanoglu committed:
Add local feature flag

alpavanoglu committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-iOS into feature/follow-conversation-tooltip

# Conflicts:
#	WordPress/Classes/ViewRelated/Feature Highlight/TooltipPresenter.swift

alpavanoglu committed:
Remove test code

alpavanoglu committed:
Add spotlight view

wargcm commented on WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:
```diff
@@ -227,6 +237,7 @@ class ReaderDetailViewController: UIViewController, ReaderDetailView {
 
         featuredImage.configure(for: post, with: self)
         toolbar.configure(for: post, in: self)
+        // Here
```
Is there supposed to be something here?

alpavanoglu replied:
Forgotten. I've deleted it.

wargcm commented on WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:
```diff
@@ -260,6 +271,61 @@ class ReaderDetailViewController: UIViewController, ReaderDetailView {
     }
 
+    private func tooltipTargetPoint() -> CGPoint? {
+        guard let followButtonMidPoint = commentsTableViewDelegate.followButtonMidPoint() else { return nil }
```
[Nitpick]: From [the style guide](https://github.com/wordpress-mobile/swift-style-guide) the `return nil` should be on the next line. Same comment applies to your other `guard` statements.

wargcm reviewed (changes requested):
Hey @alpavanoglu 👋 , I noticed a few issues while testing this:

- [The tooltip doesn't show in the correct location all the time](https://user-images.githubusercontent.com/2454408/171914679-8b24a8ce-5fc3-49af-ab16-96da8428aed2.mp4). Possibly because of images loading?
- [Switching to landscape causes the tooltip to be shown in the wrong location](https://user-images.githubusercontent.com/2454408/171914891-39ac6fb9-c1eb-4eb7-95be-290820641884.mp4).
- [On posts where there is no "Follow Conversation", it is still appearing](https://user-images.githubusercontent.com/2454408/171914971-e183a141-8a1f-4630-a876-a5902d182bc8.mp4).

---

Additionally, the PR build is failing when trying to build Jetpack:
<details>
<summary>Build log</summary>

```
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:95:13: cannot find 'FeatureHighlightStore' in scope
 
            FeatureHighlightStore.followConversationTooltipCounter += 1
                                                      ^
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:157:35: cannot find type 'TooltipPresenter' in scope
 
    private var tooltipPresenter: TooltipPresenter?
            ^~~~~~~~~~~~~~~~~~~~~
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:158:26: cannot find type 'Tooltip' in scope
 
    private var tooltip: Tooltip?
                                  ^~~~~~~~~~~~~~~~
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:285:23: cannot find 'Tooltip' in scope
 
        let tooltip = Tooltip()
                         ^~~~~~~
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:292:28: cannot find 'TooltipPresenter' in scope
 
        tooltipPresenter = TooltipPresenter(
                      ^~~~~~~
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:297:17: cannot find 'FeatureHighlightStore' in scope
 
                FeatureHighlightStore.didDismissTooltip = true
                           ^~~~~~~~~~~~~~~~
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:300:54: cannot infer contextual base in reference to member 'above'
 
        tooltipPresenter?.tooltipVerticalPosition = .above
                ^~~~~~~~~~~~~~~~~~~~~
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:434:12: cannot find 'FeatureHighlightStore' in scope
 
        && FeatureHighlightStore.shouldShowTooltip
                                                    ~^~~~~
 
 
❌  /usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailViewController.swift:1132:29: type of expression is ambiguous without more context
 
        guard let tooltip = tooltip else { return }
```
</details>

It looks like the tooltip files need to add Jetpack as a target.

alpavanoglu committed:
Make `maxWidth` and `maxContentWidth` non-static

alpavanoglu committed:
Add in & out animation on orientation change

alpavanoglu committed:
Fix arrow position for landscape orientation

alpavanoglu committed:
Fix spotlight offSetX

alpavanoglu committed:
Remove `tooltip` instance from `ReaderDetailViewController`

alpavanoglu committed:
Move guard-returns to new line

alpavanoglu committed:
Remove temporary test code

alpavanoglu commented:
Hey @wargcm . I was already working on the orientation change issue and it has proven to be a tricky issue. I added a fix for it along with the pulse view but it grew the PR in size. Apologies for that. 

I must have addressed the points you've raised apart from the first one. I can't really reproduce it. Is there anything specifically you do different or anything different with the posts you view? Every post works fine for me so any insight to reproduce it would be very helpful. Thank!

wargcm commented:
@alpavanoglu It might have something to do with it being the very first post I open on a fresh install/session. [I was able to get it to happen with your latest changes as well](https://user-images.githubusercontent.com/2454408/172263315-014f5232-50d6-4424-aae3-c0a7f28137e8.mp4). I'm deleting the app, signing in, going to Reader > Discover, and then tapping on a random blog.

Other than that:
- ✅ The landscape issue is fixed 
- ✅ The follow appearing on a post with no 'Follow conversation' is fixed 
- ❌ It still looks like there is a build issue with Jetpack

alpavanoglu committed:
Add Tooltip files to Jetpack Target

alpavanoglu committed:
Add delay for tooltip configuration

alpavanoglu commented:
@wargcm I have added a hacky fix for the issue. I have tried everything I can think of to understand why this issue occurs on a fresh install. No luck. Whole execution order seems reasonable but the position for tooltip calculation returns wrong value.

wargcm reviewed (approved):
@alpavanoglu Yep, that hack does fix the issue. Maybe in the future, we can revisit it and find out the root cause.

Everything else is working well 👍. Thanks for addressing my feedback! :shipit:
